src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/Dimension.java

```diff
@@ -81,27 +94,164 @@ protected DomainSummary getDomainSummary(FeatureCollection features, String attr
+            Tuple<String, List<Range>> specAndBuckets = HistogramUtils.getTimeBuckets(min, max, resolutionSpec);
+            List<Range> buckets = specAndBuckets.second;
+            Range<Date> referenceBucket = buckets.get(0);
+            double resolution = referenceBucket.getMaxValue().getTime() - referenceBucket.getMinValue().getTime();
+
+            // the aggregation expression classifies results in buckets numbered from 1 on
+            Function classifier = ff.function("floor",
+                    ff.divide(
+                            ff.function("dateDifference", dimensionProperty, ff.literal(min)),
+                            ff.literal(resolution)));
+            TreeMap<Object, Object> results = groupByDomainOnExpression(filter, classifier, dimensionAttributeName,
+                    Integer.class);
+
+            // map out domain representation
+            List<Integer> counts = new ArrayList<>(buckets.size());
+            for (int i = 0; i < buckets.size(); i++) {
+                Number count = Optional.ofNullable((Number) results.get(i)).orElse(0);
+                counts.add(count.intValue());
+            }
+            return Tuple.tuple(specAndBuckets.first, counts);
+        } else {
+            // assuming custom dimension, will handle as strings, once there is support
+            // for custom dimensions of different type (for structured readers) this will
+            // have to be modified
+            TreeMap<Object, Object> results = groupByDomainOnExpression(filter, dimensionProperty, dimensionAttributeName, String.class);
+            
+            // map out domain representation and histogram value representation
+            List<Integer> counts = results.values().stream()
+                    .map(v -> ((Number) v).intValue())
+                    .collect(Collectors.toList());
+            String domainRepresentation = results.keySet().stream()
+                    .map(v -> v.toString())
+                    .collect(Collectors.joining(","));
+                    
+            return Tuple.tuple(domainRepresentation, counts);
+        }
     }
 
+    /**
+     * Should we load data in memory to compute histograms (fast for small datasets not having
+     * indexes) or do we try to use visitor and perform one or more data scans instead?
+     * Small easter egg to allow testing, we might want to extend it to a per layer configuration
+     * in case there are large shapefile layers involved in this (computing min/max/groupby is
+     * three full data scans in that case). Or have a way to figure out if a collection can optimize
+     * out a visit (completely missing right now, that would be a significant API change).
+     * @return
+     */
+    private boolean loadDataInMemory() {
+        String value = Optional.ofNullable(GeoServerExtensions.getProperty("WMTS_HISTOGRAM_IN_MEMORY"))
+                .orElse("false");
+        return Boolean.getBoolean(value);
```

Comment:
If ``value`` is ``NULL``, ``Boolean.getBoolean(value)`` will return ``FALSE`` anyway, but I understand that this is more explicit.

Reply:
Good point, I can remove that optional